contracts/staking/VVVLaunchpadStaking.sol

```diff
@@ -9,22 +9,135 @@ import { VVVAuthorizationRegistryChecker } from "contracts/auth/VVVAuthorization
  */
 
 contract VVVLaunchpadStaking is VVVAuthorizationRegistryChecker {
+    ///@notice denominator for calculating early unstake penalties
+    uint256 public constant PENALTY_DENOMINATOR = 1e18;
+
+    ///@notice contains details for each stake action made by a user
+    struct Stake {
+        uint256 durationIndex;
+        uint256 amount;
+        uint256 startTimestamp;
+    }
+
+    ///@notice numerator for calculating early unstake penalties
+    uint256 public penaltyNumerator = 5e17;
+
     ///@notice the array of staking durations for each pool
     uint256[] public stakingDurations;
 
+    ///@notice the array of all stakes
+    Stake[] public allStakes;
+
+    ///@notice maps users and pools (durations) to their stake details
+    mapping(address => mapping(uint256 => Stake)) public userStakes;
```

Comment:
I like the idea of avoiding the redundant write. It could work as long as it contains all stakers past and present, so the `unstake` call doesn't need to remove it if the user unstakes all active stakes / all iteration is in the external view functions.

However, after playing around with reading all stakes in various ways, this is making me realize it could save some headache to index txns and store the array of all stakes off-chain. Not sure how large this array could get / how often it'd have to be retrieved, but I'm gathering that a max of 20k stakes (4 per each of 5000 nodes) could rival the RPC max response size and cause problems there. At any rate it seems like a lot of redundant retrieval.

Easy enough to divert that functionality?

Reply:
Sounds good 👍🏼